design-proposals/cpu-hotplug.md

```diff
@@ -0,0 +1,146 @@
+
+The user should be able to inspect the VM.Status values below to determine the state of their changes. This is similar in concept to someone inspecting a Deployments status to determine the desired number of pods vs the number of pods that are available.
+
+```
+// The declared number of sockets the vm is converging on
+vmi.Status.Resources.Sockets.Desired
+vm.Status.Resources.Sockets.Desired
+
+// The number of sockets the vm has currently active.
+vmi.Status.Resources.Sockets.Ready
+vm.Status.Resources.Sockets.Ready
+```
+
+Conditions on the VM and VMI should indicate the status of the pending CPU hotplug actions. When errors occur, the conditions should propagate the errors to the VM.Status.Conditions field so users can receive feedback into the status of their declared state.
+
+```
+// VirtualMachineCPUReady indicates that the desired amount of cpu is active on the vmi
+VirtualMachineCPUReady VirtualMachineConditionType = "CPUReady"
+```
+
+When the desired state is equal to the ready state, then these conditions will have `Status=True`, otherwise `Status=False` with a human readable reason and message indicating whether the hotplug change is pending or encountering errors.
+
+## Extensibility
+We will be able to use the new `liveUpdateFeatures` section to manage things like memory hotplug in the future.
+
+## Update/Rollback Compatibility
+This is an entirely new feature, so updates from older versions will not be a problem.
+
+## Functional Testing Approach
+Functional tests will simply:
+- Create and start a VM with `cpu` set as one of the `liveUpdateFeatures`
+- Modify the number of CPU sockets for the VM
+- Ensure the guest sees the changed number of CPUs
+- Ensure the CPU resources on the virt-launcher pod were adjusted accordingly
+
+# Implementation Phases
+
+The first two phases will happen together in the same pull request (PR), since the first phase is mostly useless without the second one.
+In the unlikely case that hot-unplugging CPUs is more complicated than anticipated, hotplug-only could be submitted as a first PR, then hot-unplug later on as a second PR.
+The third phase will happen later as a separate PR. It should be able to transparently replace the migration mechanism without API changes.
+
+## Phase 1: Creating the KubeVirt API and linking it to the LibVirt CPU hotplug API
+## Phase 2: Triggering a live migration when the number of sockets changes, to adjust the virt-launcher CPU resources
+
+The VMI [workload update controller](https://github.com/kubevirt/kubevirt/blob/ce8aab7874d2e1586787e3e2a17306b7edca1b8a/pkg/virt-controller/watch/workload-updater/workload-updater.go#L476) is extended to live migrate VMIs to satisfy updates to a VMI resource request/limits field.
+
+Sequence of events:
+- User updates CPU sockets on a vm.spec.template.spec.domain.cpu
+- VM controller determines the request/limits required and writes the necessary changes to the active vmi.Spec
```

Comment:
I probably need to rework this sentence. The idea is to disallow user specified CPU requests/limits when the CPU live update feature is enabled, because we do need to dynamically change the CPU request values. If we didn't do that, the virt-launcher pod won't get more CPU resources, and increasing the number of CPU sockets will have little to no effect...